ansible/playbooks/roles/grafana/tasks/upgrade/main.yml

```diff
@@ -0,0 +1,97 @@
+---
+- name: Include new version of defaults
+  include_vars:
+    file: defaults/versions.yml
+    name: grafana_versions
+
+- name: Include upgrade defaults
+  include_vars:
+    file: defaults/upgrade.yml
+    name: upgrade_defaults
+
+- name: Populate service facts
+  service_facts:
+  when: ansible_facts.service is undefined
+
+# If state file exists it means the previous run failed
+- name: Check for upgrade state file
+  stat:
+    path: "{{ upgrade_defaults.state_file_path }}"
+    get_attributes: false
+    get_checksum: false
+    get_mime: false
+  register: state_file_status
+
+- name: Create upgrade state file
```

Comment:
I would move this task to a place where we already know that upgrade is required (after `Check versions and set grafana_will_upgrade` task). If upgrade is already performed (rerun) there is no need to create the state file since upgrade is skipped.